File: Bayesian Inference Tools/3. Improved TMCMC/iTMCMC_Matlab/main_example_2_multimodal.m
%% example 2: multi-modal posterior
%{
---------------------------------------------------------------------------
By:                Date:          Topic:
Fong-Lin Wu        June.2019      Complete preliminary code
---------------------------------------------------------------------------
Current version 2021-03
* use of log-evidence logcE
---------------------------------------------------------------------------
References:
1."Asymptotically independent Markov sampling: a new MCMC scheme for Bayesian inference"
   James L. Beck and Konstantin M. Zuev
   International Journal for Uncertainty Quantification, 3.5 (2013) 445-474.
---------------------------------------------------------------------------
%}
clear; clc; close all;

% add path to ERADist and ERANataf classes
% Source: https://www.bgu.tum.de/era/software/eradist/
% addpath('../../../ERA_Dist/ERADistNataf_MATLAB/')

%% initial data
d = 2;      % number of dimensions (number of uncertain parameters)

%% prior PDF for theta (bivariate uniform)
a     = 10;
bound = repmat([0 a],d,1);   % boundaries of the uniform distribution box

% define the prior
prior_pdf(1) = ERADist('uniform','PAR',[bound(1,1),bound(1,2)]);
prior_pdf(2) = ERADist('uniform','PAR',[bound(2,1),bound(2,2)]);

% correlation matrix
R = eye(d);   % independent case

% object with distribution information
prior_pdf = ERANataf(prior_pdf,R);

%% likelihood PDF (mixture of gaussians in 2D)
s = RandStream.create('mt19937ar','seed',12001);   % fixed the seed to get the same data points
RandStream.setGlobalStream(s);
%
M     = 10;                          % number of mixture
mu    = prior_pdf.random(M);        % Here using ERDist.random
sigma = 0.1;
Cov   = (sigma^2)*eye(d);
w     = 0.1*ones(M,1);
%
% likelihood     = @(theta) sum(w.*mvnpdf(theta,mu,Cov)); 
likelihood     = @(theta) sum(w.*mvnpdf(repmat(theta,M,1),mu,Cov)); 
log_likelihood = @(theta) log(likelihood(theta)+realmin); 

%% iTMCMC step
Ns = 1e3;        % number of samples per level
Nb = 0.1*Ns;     % burn-in period
rng('shuffle')
fprintf('\niTMCMC: \n');
[samplesU, samplesX, q, logcE] = iTMCMC(Ns, Nb, log_likelihood, prior_pdf);

%% plot samples
m = length(q);   % number of stages (intermediate levels)

% show results
fprintf('\nModel evidence = %g', exp(logcE));
fprintf('\nMean value of x_1 = %g', mean(samplesX{m}(1,:)));
fprintf('\nStd of x_1 = %g', std(samplesX{m}(1,:)));
fprintf('\nMean value of x_2 = %g', mean(samplesX{m}(2,:)));
fprintf('\nStd of x_2 = %g\n\n', std(samplesX{m}(2,:)));

% plot q values
figure;
plot(0:m-1,q,'ro-');
xlabel('Intermediate levels $j$','Interpreter','Latex','FontSize', 18);
ylabel('$q_j$','Interpreter','Latex','FontSize', 18);
set(gca,'FontSize',15); axis tight;

% plot samples increasing q
idx = [1, round(m/3), round(2*m/3), m];
figure;
for i = 1:4
    subplot(2,2,i); plot(samplesU{idx(i)}(:,1),samplesU{idx(i)}(:,2),'r.');
    title(sprintf('$q_j$=%4.3f',q(idx(i))),'Interpreter','Latex','FontSize', 18);
    xlabel('$u_1$','Interpreter','Latex','FontSize', 18);
    ylabel('$u_2$','Interpreter','Latex','FontSize', 18);
    set(gca,'FontSize',15);  axis equal;
end
annotation('textbox', [0, 0.9, 1, 0.1],'String', '\bf Standard space', ...
    'EdgeColor', 'none', 'HorizontalAlignment', 'center');

figure;
for i = 1:4
    subplot(2,2,i); plot(samplesX{idx(i)}(:,1),samplesX{idx(i)}(:,2),'b.');
    title(sprintf('$q_j$=%4.3f',q(idx(i))),'Interpreter','Latex','FontSize', 18);
    xlabel('$\theta_1$','Interpreter','Latex','FontSize', 18);
    ylabel('$\theta_2$','Interpreter','Latex','FontSize', 18);
    set(gca,'FontSize',15);  axis equal;
end
annotation('textbox', [0, 0.9, 1, 0.1],'String', '\bf Original space', ...
    'EdgeColor', 'none', 'HorizontalAlignment', 'center');

File: Bayesian Inference Tools/3. Improved TMCMC/iTMCMC_Python/main_example_2_multimodal.py
"""
---------------------------------------------------------------------------
Example 2: multi-modal posterior
---------------------------------------------------------------------------
Created by:
Fong-Lin Wu   
Felipe Uribe
Luca Sardi

Engineering Risk Analysis Group
Technische Universitat Munchen
www.bgu.tum.de/era
---------------------------------------------------------------------------
Last Version 2021-03
* Adaptation to new ERANataf class
---------------------------------------------------------------------------
References:
1."Transitional Markov Chain Monte Carlo: Observations and improvements".
   Wolfgang Betz et al.
   Journal of Engineering Mechanics. 142.5 (2016) 04016016.1-10
2."Asymptotically independent Markov sampling: a new MCMC scheme for Bayesian inference"
   James L. Beck and Konstantin M. Zuev
   International Journal for Uncertainty Quantification, 3.5 (2013) 445-474.
---------------------------------------------------------------------------
"""

#=================================================================
# import libraries
import numpy as np
import numpy.matlib
import scipy as sp
from scipy.stats import multivariate_normal as mvn
import matplotlib.pylab as plt

# Make sure ERADist, ERANataf classes are in the path
# https://www.bgu.tum.de/era/software/eradist/
from ERADist import ERADist
from ERANataf import ERANataf
from iTMCMC import iTMCMC
plt.close('all')
np.random.seed(2021)   # fix seed

#=================================================================
# define the prior (bivariate uniform)
d     = 2      # number of dimensions (uncertain parameters)
a     = 10
bound = np.matlib.repmat([0, a],d,1)   # boundaries of the uniform box

# define the prior
prior_pdf_1 = ERADist('uniform','PAR',[bound[0,0],bound[0,1]])
prior_pdf_2 = ERADist('uniform','PAR',[bound[1,0],bound[1,1]])
prior_pdf   = [prior_pdf_1, prior_pdf_2]

# correlation matrix
R = np.eye(d)   # independent case

# object with distribution information
prior_pdf = ERANataf(prior_pdf,R)

#=================================================================
# define likelihood
# data set
m     = 10                          # number of data points
mu    = prior_pdf.random(m)         # Here using ERANataf.random
Sigma = (0.1**2)*np.eye(d)
w     = 0.1*np.ones(m)

# define likelihood
realmin        = np.finfo(np.double).tiny  # to prevent log(0)
likelihood     = lambda theta: sum(w[i] * mvn.pdf(theta, mu[i], Sigma) for i in range(0,m))
log_likelihood = lambda theta: np.log(likelihood(theta.T) + realmin)

#=================================================================
# iTMCMC step
Ns = int(1e3)      # number of samples per level
Nb = int(0.1*Ns)   # burn-in period

print('\n\niTMCMC: ')
samplesU, samplesX, q, logcE = iTMCMC(Ns, Nb, log_likelihood, prior_pdf)

#=================================================================
m = len(q)   # number of stages (intermediate levels)

# show results
print('\nModel evidence =', np.exp(logcE),'\n')
print('Mean value of x_1 =', np.mean(samplesX[-1][:,0]))
print('Std of x_1 =', np.std(samplesX[-1][:,0]),'\n')
print('Mean value of x_2 =', np.mean(samplesX[-1][:,1])) 
print('Std of x_2 =', np.std(samplesX[-1][:,1]))

#=================================================================
# Plots
# Options for font-family and font-size
plt.rc('font', size=12)
plt.rc('axes', titlesize=20)    # fontsize of the axes title
plt.rc('axes', labelsize=18)    # fontsize of the x and y labels
plt.rc('figure', titlesize=20)  # fontsize of the figure title

# plot q values
plt.figure()
plt.plot(np.arange(0,m),q,'ro-')
plt.xlabel(r'Intermediate levels $j$') 
plt.ylabel(r'$q_j$')
plt.tight_layout()
   
# plot samples increasing q
idx = np.array([0, int(np.round((m-1)/3)), int(np.round(2*(m-1)/3)), m-1])
plt.figure()
for i in range(4):
   plt.subplot(2,2,i+1) 
   plt.plot(samplesX[idx[i]][:,0], samplesX[idx[i]][:,1],'b.')
   plt.title(r'$p_j=' + str(np.round(q[idx[i]],5)) + r'$')
   plt.xlabel(r'$x_1$')
   plt.ylabel(r'$x_2$')
plt.tight_layout()

plt.show()